react/features/mobile/watchos/middleware.js

```diff
@@ -0,0 +1,203 @@
+ */
+watchOSEnabled && MiddlewareRegistry.register(store => next => action => {
+    switch (action.type) {
+    case APP_WILL_MOUNT:
+        _appWillMount(store);
+        break;
+    case CONFERENCE_JOINED:
+        store.dispatch(setConferenceTimestamp(new Date().getTime()));
+        _updateApplicationContext(store.getState());
+        break;
+    }
+
+    return next(action);
+});
+
+/**
+ * Registers listeners to the react-native-watch-connectivity lib.
+ *
+ * @param {Store} store - The redux store.
+ * @private
+ * @returns {void}
+ */
+function _appWillMount({ dispatch, getState }) {
+
+    watch.subscribeToWatchState((error, watchState) => {
+        if (error) {
+            logger.error('Error getting watchState', error);
+
+            return;
+        }
+
+        if (watchState.toLowerCase() === 'activated') {
+            _updateApplicationContext(getState);
+        }
+    });
+
+    watch.subscribeToMessages((error, message) => {
+        if (error) {
+            logger.error('watch.subscribeToMessages error:', error);
+
+            return;
+        }
+
+        const {
+            command,
+            sessionID
+        } = message;
+        const currentSessionID = _getSessionId(getState());
+
+        if (!sessionID || sessionID !== currentSessionID) {
+            logger.warn(
+                `Ignoring outdated watch command: ${message.command}`
+                    + ` sessionID: ${sessionID} current session ID: ${currentSessionID}`);
+
+            return;
+        }
+
+        switch (command) {
+        case 'joinConference': {
+            const newConferenceURL = message.data;
+            const oldConferenceURL = _getCurrentConferenceUrl(getState());
+
+            if (oldConferenceURL !== newConferenceURL) {
+                dispatch(appNavigate(newConferenceURL));
+            }
+            break;
+        }
+        case 'setMuted':
+            dispatch(
+                setAudioMuted(
+                    message.muted === 'true',
+                    /* ensureTrack */ true));
+            break;
+        case 'hangup':
+            if (_getCurrentConferenceUrl(getState()) !== 'NULL') {
+                dispatch(appNavigate(undefined));
+            }
+            break;
+        }
+    });
+}
+
+/**
+ * Figures out what's the current conference URL which is supposed to indicate what conference is currently active.
+ * When not currently in any conference and not trying to join any then the 'NULL' string value is returned.
+ *
+ * @param {Object|Function} stateful - Either the whole Redux state object or the Redux store's {@code getState} method.
+ * @returns {string}
+ * @private
+ */
+function _getCurrentConferenceUrl(stateful) {
+    const state = toState(stateful);
+    let currentUrl;
+
+    if (isInviteURLReady(state)) {
+        currentUrl = toURLString(getInviteURL(state));
+    }
+
+    // Check if the URL doesn't end with a slash
+    if (currentUrl && currentUrl.substr(-1) === '/') {
```

Comment:
we can avoid this check if we first check if there is an actual conference, right?

Reply:
Not really. It would not be a check for what this logic cares about which is if the app being in or trying to join a conference URL. When the config is being loaded there's no conference yet, but we are trying to join one. Ideally we would have a reliable way in a form of a function somewhere in `base/app` which tells that. Maybe I could try to make sure that the `getInviteURL` does that and then will remove this check.

Reply:
Kewl, please look into this. SET_ROOM happens really early, so if there is no room, there is no URL, for example.

Reply:
It is set early, but too late, because after the config will load is dispatched. I started looking into this, but it looks messy...

Reply:
this is what I need to make it work correctly:
https://github.com/jitsi/jitsi-meet/pull/4075/commits/d05c71e88805cdcc21f7899e8c9e978b52475913

Reply:
yeah, it looks a bit messy, let's stick to what was before, maybe we find a better way in the future. thanks for giving it a shot!